requests-for-comments/2022-06-10-nail-exports.md

```diff
@@ -0,0 +1,114 @@
+|------------|--------------|----------|-----------|-----------|----------|
+
+Similar to SharedDashboards we currently have, it makes sense to have a unique rendering method for standalone visualisations that can either be accessed:
+1. Directly via an authenticated user
+2. Directly via an unauthenticated user if "sharing" has been enabled
+3. Via an iframe embedded elsewhere
+4. Server side via selenium for taking high-quality image exports as "screenshots"
+
+---
+#### Technical considerations
+- [ ] Embeds should allow customiseable options as query params for showing hiding different elements (e.g. readonly, whitelabel)
+
+#### Current State
+- 🤔 Dashboards are embeddable and shareable under a special `shared_dashboards` endpoint
+- 🤔 Exports are _almost_ embeddable via the `exporter` endpoint. Currently this is only used for image rendering
+- **PROPOSAL**: Merge these endpoints together to a unified endpoint for all standalone viewing of content
+
+---
+### Images
+| insights ✅ | dashboards ✅ | events ❌ | person ❌ | cohorts ❌ | groups ❌ |
+|------------|--------------|----------|-----------|-----------|----------|
+
+Images are important either for the user (e.g. to include in a powerpoint presentation) or for use by other export methods such as subscriptions
+
+#### Technical considerations
+- [ ] Some things are hard/impossible to render nicely in small sizes (e.g. Funnels).
+- [ ] 
+
+
+---
+### Subscriptions - email:
+| insights ✅ | dashboards ✅ | events ❌ | person ❌ | cohorts ❌ | groups ❌ |
+|------------|--------------|----------|-----------|-----------|----------|
+
+Subscriptions via email are essentially Image exports but delivered via email
+
+#### Technical considerations
+- [ ] The exported Image needs to be publicly accessible using some sort of token
+- [ ] Users need to be able to easily unsubscribe without needing to login
+
+---
+### Subscriptions - slack:
+| insights ✅ | dashboards ✅ | events ❌ | person ❌ | cohorts ❌ | groups ❌ |
+|------------|--------------|----------|-----------|-----------|----------|
+
+Subscriptions via slack are essentially Image exports but delivered via slack. We may want to consider additional information to be included
+
+#### Technical considerations
+- [ ] The exported Image needs to be publicly accessible using some sort of token
+- [ ] Users need to be able to easily unsubscribe without needing to login
+
+
+---
+### CSV / Excel:
+| insights ✅ | dashboards ✅ | events ✅ | person ✅ | cohorts ✅ | groups ✅ |
+|------------|--------------|----------|-----------|-----------|----------|
+
+Resources can be exported directly to a CSV. This may be commonly done adhoc (e.g. from some in-memory filter configured on the frontend.
+
+#### Technical considerations
+- [ ] Exporting 10000s of rows can be costly or even impossible syncronously - it needs to be possible to do this in the background and send to the user when complete
```

Comment:
I would opt for a default behavior - any csv gets sent via email. Easier to implement, more unified experience regardless of csv size.

Reply:
Downside of this would be:
1. Self-hosted doesn't always have email configured
2. Alot of csv exports _will_ be small enough for a quick easy export. I'm thinking we already have the async flow for image exports so it wouldn't be hard to manage it that way as well.

Maybe if the export takes so long that an email is required, we should reconsider if we even support it...?

Reply:
I'm imagining

* someone clicks to export
* we tell them we'll generate it and let them know
* and in that toast link them to a page where they can see their historical exports
* they'd have status like "expired" e.g. we deleted it but you can regenerate it
* and have download links
* then we can also email people to let them know a download is available if email is configured 
* but they have an onward journey if it isn't